datums_warehouse/broker/source.py

```diff
@@ -1,22 +1,126 @@
+
+        return res
+
+    def _get_cached(self, since, until):
+        if not self._cache.exists():
+            return None
+
+        with gzip.open(self._cache, mode='rb') as file:
+            combined = None
+            start = time.time()
+            for line in file:
+                trades = json.loads(line, encoding='utf-8')
+                if get_last(trades) >= since:
+                    combined = combine_trades(combined, trades)
+                if get_last(combined) >= until or get_len(combined) >= self._max_results:
+                    break
+            duration = time.time() - start
+            logger.debug(f"[performance] cache lookup took: {duration}s")
+
+        logger.info(f" <<< cached, pair={self._pair}, since={since}, until={until}")
+        return combined
+
+    def _query_remote_trades(self, since):
+        logger.info(f" >>> querying {self._TRADE_URL}, pair={self._pair}, since={since}")
+        res = requests.get(self._TRADE_URL, params=dict(pair=self._pair, since=since)).json()
+        return res
+
+    def _update_cache(self, trades):
+        logger.info(f" >>> cache update, pair={self._pair}, len={get_len(trades)}")
+        self._cache.parent.mkdir(parents=True, exist_ok=True)
+        with gzip.open(self._cache, mode='ab') as file:
+            file.write(f"{json.dumps(trades)}\n".encode())
+
+
+def get_last(trades):
+    if trades is None:
+        return 0
+    return int(trades['result']['last'])
+
+
+def get_len(trades):
+    if trades is None:
+        return 0
+    pair = get_pair(trades)
+    return len(trades['result'][pair])
+
+
+def combine_trades(lhs, rhs):
+    if lhs is None:
+        return rhs
+    if rhs is None:
+        return lhs
+
+    lhs['result']['last'] = str(get_last(rhs))
+    pair = get_pair(lhs)
+    lhs['result'][pair] += rhs['result'][pair]
+    return lhs
+
+
+def get_pair(trades):
+    return first([k for k in trades['result'].keys() if k != 'last'])
+
+
+class KrakenSource:
+    def __init__(self, trades_storage, pair, interval, max_results=1e6):
+        self._trades = KrakenTrades(trades_storage, pair, max_results)
         self._interval = interval
         self._adapter = KrakenAdapter(self._interval)
+        self._server_time = KrakenServerTime()
 
-    def query(self, since):
-        res = requests.get(self._TRADE_URL, params=dict(pair=self._pair, since=to_nano_sec(since)))
-        return validate(CsvDatums(self._interval, self._adapter(res.json())))
+    def query(self, since, exclude_outliers=None, z_score_threshold=10):
+        last_itv = to_nano_sec(floor_to_interval(self._server_time.now(), self._interval * 60))
+        res = self._trades.get(to_nano_sec(since), last_itv)
+        datums = CsvDatums(self._interval, self._adapter(res))
+        try:
+            validate(datums, exclude_outliers, z_score_threshold)
+        except DataError as e:
+            logger.warning("invalid data found:\n", e.args[0])
```

Comment:
Maybe just the message here  - the call stack is not that important for the DataError